### PR TITLE
Introduce Signer as a richer API for signing domain objects

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
The `MessageSignerService` interface works purely on the signing root value leaving the caller to calculate the signing root for the specific object.  To avoid duplicating that logic and provide a richer domain-object based API for signing, introduce the `Signer` object which takes a domain object, calculates the signing root then uses the `MessageSignerService` to sign it.

Loading of validators is also adjusted to move it in the direction required for loading into the validator client service which won't know the validator index.

### CODE REVIEW

File: validator/client/src/main/java/tech/pegasys/artemis/validator/client/signer/Signer.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.operations.AttestationData;
+import tech.pegasys.artemis.datastructures.state.Fork;
+import tech.pegasys.artemis.datastructures.validator.MessageSignerService;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.bls.BLSSignature;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class Signer {
```
Comment (cemozerr) [inline validator/client/src/main/java/tech/pegasys/artemis/validator/client/signer/Signer.java:31]: Really nice and clean that all the signing is now contained in one object.

File: validator/client/src/test/java/tech/pegasys/artemis/validator/client/signer/SignerTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Test
+  public void shouldCreateRandaoReveal() {
+    final BLSSignature signature = dataStructureUtil.randomSignature();
+    final Bytes expectedSigningRoot =
+        Bytes.fromHexString("0x84c8825ccb169a0d1109ff2b96185ea2bd6f66ed198f2e322473773fff3ed91b");
+    when(signerService.signRandaoReveal(expectedSigningRoot))
+        .thenReturn(SafeFuture.completedFuture(signature));
+    final SafeFuture<BLSSignature> reveal =
+        signer.createRandaoReveal(UnsignedLong.valueOf(7), fork);
```
Comment (cemozerr) [inline validator/client/src/test/java/tech/pegasys/artemis/validator/client/signer/SignerTest.java:48]: I don't have a suggestion, however, it took me a good amount of time to infer (still not sure) that passing the epoch number 7 and the given fork value results in the expectedSigningRoot that you declare above. I wish there was a more clear way of showing that.

**Review by cemozerr — APPROVED** [review]
LGTM. Made one note on test readability. The encapsulation of signing logic is really nice.

File: validator/client/src/test/java/tech/pegasys/artemis/validator/client/signer/SignerTest.java (hunk 2 above)
Comment (ajsutton) [reply validator/client/src/test/java/tech/pegasys/artemis/validator/client/signer/SignerTest.java:48]: Yeah, ultimately those tests are really checking that the signing root was calculated correctly and that it was delegated to the right method on the `MessageSignerService`. I can't actually see a way to simplify them either.  We could have a separate object responsible for calculating the signing root and test that separately but it just makes the production code more complex and I don't think it's worth it.

**Pull request merged**